src/gurt/dyn_hash.c

```diff
@@ -0,0 +1,1269 @@
+	for (idx = 0; idx < htable->ht_vector.counter; idx++) {
+		bucket = htable->ht_vector.data[idx];
+		if (bucket == prev) {
+			/**/
+			continue;
+		}
+		prev = bucket;
+		if (!(gtable->ht_feats & D_HASH_FT_NOLOCK)) {
+			/**/
+			D_MUTEX_DESTROY(&bucket->mtx);
+		}
+		D_FREE(bucket);
+	}
+	vec_destroy(&htable->ht_vector);
+	htable->ht_rw_unlock(htable);
+
+	if (!(gtable->ht_feats & D_HASH_FT_NOLOCK)) {
+		/**/
+		if (gtable->ht_feats & D_HASH_FT_MUTEX) {
+			/**/
+			D_MUTEX_DESTROY(&gtable->ht_lock.mutex);
+		} else if (gtable->ht_feats & D_HASH_FT_RWLOCK) {
+			/**/
+			D_RWLOCK_DESTROY(&gtable->ht_lock.rwlock);
+		} else {
+			/**/
+			D_SPIN_DESTROY(&gtable->ht_lock.spin);
+		}
+	}
+
+	htable->ht_magic = 0;
+out:
+	return rc;
+}
+
+dh_item_t
+dyn_hash_rec_find(struct d_hash_table *gtable, const void *key,
+		  unsigned int ksize, uint64_t siphash)
+{
+	int		rc = 0;
+	dh_bucket_t	*bucket;
+	dh_item_t	item = NULL;
+	bool		sip_update = false;
+	uint32_t        index;
+	struct dyn_hash *htable = gtable->dyn_hash;
+
+	D_ASSERT(gtable->ht_feats & D_HASH_FT_DYNAMIC);
+	D_ASSERT(htable->ht_magic == DYNHASH_MAGIC);
+	if (siphash == 0) {
+		siphash = gen_siphash(key, ksize);
+		sip_update = true;
+	}
+	htable->ht_read_lock(htable);
+	index = (uint32_t)(siphash >> htable->ht_shift);
+	bucket = htable->ht_vector.data[index];
+	htable->bucket_lock(bucket);
+	htable->ht_rw_unlock(htable);
+
+	rc = find_exact_match(htable, bucket, siphash, key, ksize);
+	if (rc >= 0) {
+		/* get record */
+		item = bucket->field[rc].record;
+	}
+	if (item != NULL) {
+		htable->ht_ops.hop_rec_addref(gtable, item);
+		if (sip_update) {
+			/* call user operation */
+			htable->ht_ops.hop_siphash_set(item, siphash);
+		}
+	}
+	htable->bucket_unlock(bucket);
+	return item;
+}
+
+dh_item_t
+dyn_hash_rec_find_insert(struct d_hash_table *gtable, const void *key,
+			 unsigned int ksize, dh_item_t item, uint64_t siphash)
+{
+	int		rc = 0;
+	bool		set_siphash = false;
+	dh_item_t	tmp = item;
+	struct dyn_hash *htable = gtable->dyn_hash;
+
+	D_ASSERT(gtable->ht_feats & D_HASH_FT_DYNAMIC);
+	D_ASSERT(htable->ht_magic == DYNHASH_MAGIC);
+	if (siphash == 0) {
+		siphash = gen_siphash(key, ksize);
+		set_siphash = true;
+	}
+	do {
+		rc = do_insert(htable, key, ksize, &tmp, siphash,
+			       DH_LOOKUP_INSERT);
+	} while (rc == -DER_AGAIN);
+	if (rc == 0 && set_siphash) {
+		/* call user operation */
+		htable->ht_ops.hop_siphash_set(tmp, siphash);
+	}
+	return tmp;
+}
+
+int
+dyn_hash_rec_insert(struct d_hash_table *gtable, const void *key,
+		    unsigned int ksize, dh_item_t item, bool exclusive)
+{
+	int		rc = 0;
+	uint64_t	siphash;
+	struct dyn_hash	*htable = gtable->dyn_hash;
+
+	D_ASSERT(gtable->ht_feats & D_HASH_FT_DYNAMIC);
+	D_ASSERT(htable->ht_magic == DYNHASH_MAGIC);
+
+	siphash = gen_siphash(key, ksize);
+	do {
+		rc = do_insert(htable, key, ksize, &item,
+			       siphash, exclusive ? DH_INSERT_EXCLUSIVE :
+			       DH_INSERT_INCLUSIVE);
+	} while (rc == -DER_AGAIN);
```

Comment:
in both places hwere do_insert() is called we have similar loop to retry on -DER_AGAIN. it would be better for do_insert() or its wrapper to handle this logic instead